### PR TITLE
Reworked expansion

### LINKED ISSUE #13: [Request] New placeholder for showing last damage given.
This is a suggestion from https://github.com/PlaceholderAPI/PlaceholderAPI/issues/56

Idea would be to add a placeholder to show the last damage a player has given to an entity.

My idea there is to change `%player_last_damage%` by splitting it up into `%player_last_damage_taken%` for damage the player took and `%player_last_damage_given%` (or dealt?) that shows the last damage the player dealt to a entity.
Only issue I see is to (temporary) store the info, since it doesn't seem like there's a way the damage a player has given is stored somewhere to easy access.

### LINKED ISSUE #64: %player_item_in_hand_lore%
Would it be possible to add `%player_item_in_hand_lore%`, functioning exactly the same as `%player_item_in_hand_name%`, but returning the item's Lore instead?

### LINKED ISSUE #85: add new color for ping
ping 0 to 20 milliseconds color blue

### PR DESCRIPTION
**Pull request by Tanguygab**
Lots of changes 😅

Moved to Gradle
Added `%player_health_full(_rounded)%`, `%player_time_since_last_played(/join)%`, `%player_time_since_join%`, `%player_item_in_hand_lore%` with its offhand and armor variants (closes #64) and `%player_last_damage_given%` (closes #13, can't believe it took 4 years just for that)
Since I'm using enhanced switch statements now (lots of them :D) I'm guessing this now requires at least Java 14, though most servers should have updated to 17 now and other expansions bumped their Java version as well so this should be fine.

Reworked `%player_colored_ping%`'s config to use ranges of numbers to make it much more customizable, default config:
```yml
expansions:
  player:
    ping:
      '-1': '&c'
      0-50: '&a'
      50-100: '&e'
      100-: '&c'
```
Closes #85 #77

Included #84, #83, #65 and #37

If anyone wants to try, I tested it but just in case
[Player-Expansion-2.1.0.zip](https://github.com/PlaceholderAPI/Player-Expansion/files/12447768/Player-Expansion-2.1.0.zip)

### CODE REVIEW

**Comment by Tanguygab** [comment]
If anyone wants to try
[Player-Expansion-2.1.0.zip](https://github.com/PlaceholderAPI/Player-Expansion/files/12447768/Player-Expansion-2.1.0.zip)

File: src/main/java/com/extendedclip/papi/expansion/player/PlayerListener.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package com.extendedclip.papi.expansion.player;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.player.PlayerJoinEvent;
+import org.bukkit.event.player.PlayerQuitEvent;
+
+public record PlayerListener(PlayerExpansion expansion) implements Listener {
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/papi/expansion/player/PlayerListener.java:10]: Use java 8, this applies to the rest of the files too.

File: build.gradle [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        name = 'papi-repo'
+        url = 'https://repo.extendedclip.com/content/repositories/placeholderapi/'
+    }
+}
+
+dependencies {
+    compileOnly 'org.spigotmc:spigot-api:1.20.1-R0.1-SNAPSHOT'
+    compileOnly 'me.clip:placeholderapi:2.11.3'
+    compileOnly 'org.projectlombok:lombok:1.18.28'
+    annotationProcessor 'org.projectlombok:lombok:1.18.28'
```
Comment (iGabyTM) [inline build.gradle:24]: Why? Please remove this.

File: src/main/java/com/extendedclip/papi/expansion/player/PlayerExpansion.java [hunk 3]
```diff
@@ -21,469 +21,336 @@
...
+    private final PlayerListener listener;
+
+    private final Map<String,String> pingColors = new HashMap<>();
+    private final Map<BlockFace,String> directions = new HashMap<>();
+    final Map<Player, Long> joinTimes = new HashMap<>();
+    final Map<Player,Double> damagesGiven = new HashMap<>();
+
+    public PlayerExpansion() {
+        defaults = new HashMap<>() {{
+            put("ping",new HashMap<>() {{
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/papi/expansion/player/PlayerExpansion.java:60]: There is ImmutableMap#Builder from guava, please avoid using anonymous classes like this.

File: src/main/java/com/extendedclip/papi/expansion/player/PlayerExpansion.java [hunk 4]
```diff
@@ -21,469 +21,336 @@
...
+                            "armor_chestplate_name","armor_chestplate_data","armor_chestplate_durability",
+                            "armor_leggings_name","armor_leggings_data","armor_leggings_durability",
+                            "armor_boots_name","armor_boots_data","armor_boots_durability" -> {
+                        if (identifier.startsWith("item_in_hand")) yield requestItem(identifier,12,versionHelper.getItemInHand(p));
+                        if (identifier.startsWith("item_in_offhand")) yield requestItem(identifier,15,p.getInventory().getItemInOffHand());
+                        if (identifier.startsWith("armor_helmet")) yield requestItem(identifier,12,p.getInventory().getHelmet());
+                        if (identifier.startsWith("armor_chestplate")) yield requestItem(identifier,16,p.getInventory().getChestplate());
+                        if (identifier.startsWith("armor_leggings")) yield requestItem(identifier,14,p.getInventory().getLeggings());
+                        if (identifier.startsWith("armor_boots")) yield requestItem(identifier,11,p.getInventory().getBoots());
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/papi/expansion/player/PlayerExpansion.java:291]: Please avoid using magic numbers (12, 15, 16 have no meaning)

File: src/main/java/com/extendedclip/papi/expansion/player/PlayerUtil.java [hunk 5]
```diff
@@ -45,94 +46,10 @@ public final class PlayerUtil {
...
-    };
-
-
-    public static int getPing(final Player player) {
-        return PLAYER_GET_PING.apply(player);
-    }
-
-    public static String getLocale(final Player player) {
-        return PLAYER_GET_LOCALE.apply(player);
+    public static String msToSToStr(double ms) {
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/papi/expansion/player/PlayerUtil.java:51]: Use real names for methods

File: src/main/java/com/extendedclip/papi/expansion/player/VersionHelper.java [hunk 6]
```diff
@@ -44,14 +75,43 @@ private static int getCurrentVersion() {
...
+            return Integer.parseInt(stringBuilder.toString());
+        } catch (NumberFormatException e) {
             throw new IllegalArgumentException("Could not retrieve server version!");
         }
+    }
 
-        return version;
+    public int getPing(Player player) {
+        if (IS_1_17_OR_NEWER) return player.getPing();
+        try {return (int) ping.get(getHandle.invoke(player));}
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/papi/expansion/player/VersionHelper.java:87]: ???

File: src/main/java/com/extendedclip/papi/expansion/player/VersionHelper.java [hunk 7]
```diff
@@ -44,14 +75,43 @@ private static int getCurrentVersion() {
...
+        return IS_1_9_OR_NEWER ? Objects.requireNonNull(p.getAttribute(Attribute.GENERIC_MAX_HEALTH)).getValue() : p.getMaxHealth();
+    }
+
+    @SuppressWarnings("deprecation")
+    public ItemStack getItemInHand(Player p) {
+        return IS_1_9_OR_NEWER ? p.getInventory().getItemInMainHand() : p.getInventory().getItemInHand();
+    }
+    @SuppressWarnings("deprecation")
+    public int getItemDamage(ItemStack item) {
+        return VERSION >= 1_13_0 ? item.getItemMeta() instanceof Damageable damageable
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/papi/expansion/player/VersionHelper.java:111]: Put this statement on multiple lines, it is confusing to read.

File: src/main/java/com/extendedclip/papi/expansion/player/PlayerUtil.java [hunk 8]
```diff
@@ -164,60 +81,35 @@ public static BlockFace getDirection(Player player) {
...
-        }
-        if (level <= 30) {
-            return (level * 5) - 38;
-        }
-        return (level * 9) - 158;
+        return level <= 15
+                ? (level << 1) + 7
+                : level <= 30
+                    ? (level * 5) - 38
+                    : (level * 9) - 158;
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/papi/expansion/player/PlayerUtil.java:112]: Revert these changes, not everything has to be done using ternary operator.

File: src/main/java/com/extendedclip/papi/expansion/player/PlayerExpansion.java [hunk 9]
```diff
@@ -21,469 +21,336 @@
...
-        // online placeholders
-        if (!player.isOnline()) {
-            return "";
-        }
+        Object output = request(player,identifier);
+        return output == null ? null
+                : output instanceof Boolean bool ?
+                    bool ? PlaceholderAPIPlugin.booleanTrue()
+                            : PlaceholderAPIPlugin.booleanFalse()
+                : String.valueOf(output);
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/papi/expansion/player/PlayerExpansion.java:119]: Again, don't use ternary operator like this.

**Review by iGabyTM — CHANGES_REQUESTED** [review]
